### ISSUE TITLE
Values are not cached in partial applications

### ISSUE THREAD

**Issue by alfonsogarciacaro** [fable-compiler/Fable#1659]
Still not sure if we want to "fix" this, but I'll write it here so I don't forget and also in case someone has an idea. Right now this is failing in Fable:

```fsharp
let mutable private x = 0

let inline foo<'T> = x <- x + 1; 5

let add x y = x + y

testCase "Don't inline values that evaluate multiple times" <| fun () ->
    let li = [1;2;3;4]
    let res = List.map (add foo) li
    equal 1 x
    equal [6;7;8;9] res
```

When run in dotnet `foo` is only run once, but in Fable is run multiple times (so the final value of `x` is 4). The issue is Fable is doing an [optimistic inlining](https://github.com/fable-compiler/Fable/blob/f606721155b8ea9447697321fecac9b5175cfa0e/src/dotnet/Fable.Compiler/Transforms/FableTransforms.fs#L202-L204) for the bindings generated by the F# compiler in situations like this. Even if the value has a risk of being evaluated multiple times, we still inline it if it's only referenced once. The problem is when we inline it within a function and the function itself is run multiple times.

I tried making the inlining more strict, but then most of the closures generated by the F# compiler (in the unoptimized AST) in pipe chaining, partial application, format printing, etc, don't get optimized and the result is a bigger and less readable JS output, so this may not be acceptable side effect to fix this (after all, nobody has complained about this so far).

Any other ideas how could this be fixed? @ncave

**Comment by ncave** [comment]
@alfonsogarciacaro I'm not sure, is it possible to not inline values at all (only inline functions)? Or is that worse?

**Comment by alfonsogarciacaro** [comment]
I guess we can do that. We just get more closures, for examples if I change `canInlineArg` to:

```fsharp
let canInlineArg identName value body =
    not(hasDoubleEvalRisk value)
        || (match value with
            | Function _ -> countReferences 1 identName body <= 1
            | _ -> false)
```

A function like this:

```fsharp
let equal expected actual =
    printfn "%A = %A > %b" expected actual (expected = actual)
```

Becomes:

```js
function equal(expected, actual) {
  var clo1$$1, clo1$$2;
  (clo1$$1 = (0, _String.toConsole)((0, _String.printf)("%A = %A > %b")), function (arg10$$1) {
    const clo2 = clo1$$1(arg10$$1);
    return function (arg20) {
      const clo3 = clo2(arg20);
      return function (arg30) {
        clo3(arg30);
      };
    };
  })(expected)(actual)((0, _Util.equals)(expected, actual));
}
```

Instead of (current result in the REPL):

```js
function equal(expected, actual) {
  toConsole(printf("%A = %A > %b"))(expected)(actual)(equals(expected, actual));
}
```

But as you always say we need to measure to see if this is an actual problem or not. I'll push a branch with the change to see how much the bundle size changes because of it :+1:

**Comment by alfonsogarciacaro** [comment]
When [making inlining more strict](https://github.com/fable-compiler/Fable/commit/a63b70f2437446d7d4499665186587bf4e871e97), we get a [REPL bundle size of 3836KB](https://ci.appveyor.com/project/fable-compiler/fable/builds/20825782#L1741) (vs 3781KB). There seems to be a performance hit as well :/

**Comment by 0x53A** [comment]
There is ``ExprHasEffect`` in the compiler to check if an expression is pure. Afaik it is relatively conservative and should not have any false-positives (but it obviously has false-negatives).

Could this be used to guide inlining?

**Comment by alfonsogarciacaro** [comment]
Ah, that could be an option! It's the `ExprHasEffect` in `Microsoft.FSharp.Compiler.Optimizer`? It'd be nice to have access to it but unfortunately right now it seems [to be internal](https://github.com/Microsoft/visualfsharp/blob/44c7e10ca432d8f245a6d8f8e0ec19ca8c72edaf/src/fsharp/Optimizer.fs#L10).

**Comment by alfonsogarciacaro** [comment]
But it's a good pointer nonetheless. Maybe we can adapt it, I already have something similar [here](https://github.com/fable-compiler/Fable/blob/4229742047f4aca43d5f26d20b9e7873f4a9c160/src/dotnet/Fable.Compiler/Transforms/Transforms.Util.fs#L264-L276).

**Comment by alfonsogarciacaro** [comment]
Fixed by a3796d0bb5c73a22413472bb538476f179354635
